opm/autodiff/WellStateFullyImplicitBlackoil.hpp

```diff
@@ -191,6 +191,51 @@ namespace Opm
         std::vector<double>& wellPotentials() { return well_potentials_; }
         const std::vector<double>& wellPotentials() const { return well_potentials_; }
 
+        data::Wells report( const Wells* wells ) const override {
+            using rt = data::Rates::opt;
+
+            data::Wells res = WellState::report( wells );
+
+            const int nw = this->numWells();
+            const int np = this->numPhases();
+
+            for( auto w = 0; w < nw; ++w ) {
+
+                std::map< size_t, data::Completion > completions;
+                const auto* begin = wells->well_connpos + w;
+                const auto* end = wells->well_connpos + w + 1;
+                for( auto* i = begin; i != end; ++i ) {
+
+                    const auto perfrate = this->perfPhaseRates().begin() + *i;
+                    data::Rates perfrates;
+                    perfrates.set( rt::wat, *(perfrate + 0) );
+                    perfrates.set( rt::oil, *(perfrate + 1) );
+                    perfrates.set( rt::gas, *(perfrate + 2) );
```

Comment:
It seems to me this is the final remaining hurdle to clear for this. The problem is that we are going from a phase-agnostic code (Wells [from opm-core] and WellState know nothing about which phases are active, only their number) to a code which explicitly must be told "this is an oil rate" etc.

The only solution I can see is to do these three steps:
 - add a PhaseUsage member to WellState
 - add as an argument in the init() function
 - at the call sites pass the proper object (for example in SimulatorBase_impl.hpp line 153 pass `props_.phaseUsage()`).

Then it is possible to both get the correct indices (into perfrate) for each phase, and to check if it is active etc.

Without, this code may segfault at any time (since the state can contain just data for two phases).